src/lib/cfa/hosts.rb

```diff
@@ -0,0 +1,189 @@
+require "yast"
+require "yast2/target_file"
+
+require "cfa/base_model"
+require "cfa/matcher"
+require "cfa/augeas_parser"
+
+module CFA
+  # class representings /etc/hosts file model. It provides helper to manipulate
+  # with file. It uses CFA framework and Augeas parser.
+  # @see http://www.rubydoc.info/github/config-files-api/config_files_api/CFA/BaseModel
+  # @see http://www.rubydoc.info/github/config-files-api/config_files_api/CFA/AugeasParser
+  class Hosts < BaseModel
+    PARSER = AugeasParser.new("hosts.lns")
+    PATH = "/etc/hosts".freeze
+    include Yast::Logger
+
+    def initialize(file_handler: nil)
+      super(PARSER, PATH, file_handler: file_handler)
+    end
+
+    # The old format used by {Yast::HostClass}.
+    # @return [Hash{String => Array<String>}] keys are IPs,
+    #   values are lists of lines in /etc/hosts (not names!)
+    #   with whitespace separated hostnames, where the first one is canonical
+    #   and the rest are aliases
+    #
+    #   For example, the file contents
+    #
+    #       1.2.3.4 www.example.org www
+    #       1.2.3.7 log.example.org log
+    #       1.2.3.7 sql.example.org sql
+    #
+    #   is returned as
+    #
+    #       {
+    #         "1.2.3.4" => "www.example.org www"
+    #         "1.2.3.7" => [
+    #           "log.example.org log",
+    #           "sql.example.org sql"
+    #         ]
+    #       }
+    def hosts
+      matcher = Matcher.new { |k, _v| k =~ /^\d*$/ }
+      data.select(matcher).each_with_object({}) do |host, result|
+        entry = host[:value]
+        result[entry["ipaddr"]] ||= []
+        result[entry["ipaddr"]] << single_host_entry(entry)
+      end
+    end
+
+    # Returns single entry from hosts for given ip or empty array if not found
+    # @see #hosts
+    # @return [Array<String>]
+    def host(ip)
+      hosts = data.select(ip_matcher(ip))
+
+      hosts.map do |host|
+        single_host_entry(host[:value])
+      end
+    end
+
+    # deletes all occurences of given ip in host table
+    # @return [void]
+    def delete_host(ip)
+      entries = data.select(ip_matcher(ip))
+      if entries.empty?
+        log.info "no entry to delete for ip #{ip}"
+        return
+      end
+
+      if entries.size > 1
+        log.info "delete host with ip '#{ip}' removes more then one entry"
+      end
+
+      entries.each do |e|
+        log.info "deleting record #{e.inspect}"
+        data.delete(e[:key])
+      end
+    end
+
+    # Replaces or adds a new host entry.
+    # If more than one entry with the given ip exists
+    # then it replaces the last instance.
```

Comment:
@mvidner do you see it intuitive? I really do not expect such behavior. Currently it is used only in one place `Host.Update` which is in fact also quite confusing. So what Update did? it removed old hostname and adds new hostname to given ip. So from my non-expert point of view I expect that in fact this set_entry is not needed and instead add_entry should be used in this case after `remove_hostname` (which is currently used ).

Also according to googling it looks like all of them are canonical and all of them will be resolved. So if you use hostA.suse.cz or hostB.suse.cz or hostC.suse.cz it should point you to same ip. And reverse it is first occurence, according to debian discussion - http://lists.debian.org/debian-devel/2004/06/msg00468.html so host for ip will get hostA.suse.cz